File: scripts/unlock-user.js
'use strict';

// Break-glass account unlock.
//
//   docker exec -it <dashboard-container> npm run unlock-user -- <username>
//   docker exec -it <dashboard-container> npm run unlock-user -- --list
//
// Account lockouts expire on their own after 30 minutes (see src/auth.js) and an
// admin can clear one from Settings → Users, so this script is not the normal
// path. It exists for the case those two cannot help: every admin account locked
// at the same time, with /setup already sealed. Without it the only way back in
// would be hand-written SQL against the production database.
//
// It deliberately CANNOT create an account or change a password. Locking is the
// only thing it touches, so possessing it is not equivalent to possessing the
// dashboard.

const db = require('../src/db');

async function main() {
  const arg = process.argv[2];

  if (!arg || arg === '--help' || arg === '-h') {
    console.log('usage: npm run unlock-user -- <username>');
    console.log('       npm run unlock-user -- --list      # show locked accounts');
    process.exit(arg ? 0 : 1);
  }

  const reachable = await db.waitForDb({ attempts: 3, delayMs: 1500 });
  if (!reachable) {
    console.error('[unlock] cannot reach PostgreSQL — check DB_HOST/DB_PORT and the db_password secret');
    process.exit(2);
  }

  if (arg === '--list') {
    const rows = await db.all(
      `SELECT username, role, failed_attempts, locked_until
         FROM users
        WHERE locked_until > NOW()
        ORDER BY username`
    );
    if (!rows.length) {
      console.log('[unlock] no accounts are currently locked');
    } else {
      console.log(`[unlock] ${rows.length} locked account(s):`);
      for (const r of rows) {
        const mins = Math.ceil((new Date(r.locked_until).getTime() - Date.now()) / 60_000);
        console.log(`  ${r.username} (${r.role}) — ${r.failed_attempts} failed, unlocks in ${mins} min`);
      }
    }
    return;
  }

  const { rowCount } = await db.query(
    'UPDATE users SET failed_attempts = 0, locked_until = NULL WHERE username = $1',
    [arg]
  );

  if (!rowCount) {
    console.error(`[unlock] no such user: ${arg}`);
    process.exit(3);
  }
  console.log(`[unlock] '${arg}' unlocked — failed attempt counter reset`);
}

main()
  .then(() => db.pool.end())
  .catch(async (err) => {
    console.error(`[unlock] failed: ${err.message}`);
    try { await db.pool.end(); } catch { /* already closing */ }
    process.exit(1);
  });

File: scripts/unlock-user.py
#!/usr/bin/env python3
"""Break-glass account unlock.

The situation this exists for: every admin account is locked at the same time.
The automatic lockout expires on its own after 30 minutes, and an admin can
clear another admin's lock from the UI — but if the *only* admin is locked and
somebody set the manual ``users.locked`` flag, there is no path back in through
the browser, because /setup sealed itself when the first account was created.

Run it on a host that can reach PostgreSQL, with the same credentials the
dashboard uses:

    DB_PASS=... python scripts/unlock-user.py --list
    DB_PASS=... python scripts/unlock-user.py --user siro
    DB_PASS=... python scripts/unlock-user.py --all

Inside the running container the password is already mounted as a Swarm secret,
so no environment variable is needed at all:

    docker exec -it <container> python scripts/unlock-user.py --list

It deliberately cannot set a password. Restoring access to a locked account is a
different operation from taking one over, and a script that could do both would
be the most dangerous file in the repository.
"""

from __future__ import annotations

import argparse
import asyncio
import sys
from pathlib import Path

# Run from a checkout as well as from /app inside the image.
sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import asyncpg  # noqa: E402

from src import config  # noqa: E402


def _connect_kwargs() -> dict:
    if config.DATABASE_URL:
        return {"dsn": config.DATABASE_URL}
    return {
        "host": config.DB_HOST,
        "port": config.DB_PORT,
        "database": config.DB_NAME,
        "user": config.DB_USER,
        "password": config.DB_PASS,
    }


async def main() -> int:
    parser = argparse.ArgumentParser(description="Clear dashboard account lockouts.")
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--list", action="store_true", help="show every account and its lock state")
    group.add_argument("--user", metavar="USERNAME", help="unlock one account")
    group.add_argument("--all", action="store_true", help="unlock every locked account")
    args = parser.parse_args()

    print(f"[unlock] target: {config.describe_db_target()}")
    try:
        con = await asyncpg.connect(**_connect_kwargs())
    except Exception as err:  # noqa: BLE001
        print(f"[unlock] cannot connect: {err}", file=sys.stderr)
        return 2

    try:
        if args.list:
            rows = await con.fetch(
                """SELECT id, username, role, failed_attempts, locked,
                          CASE WHEN locked_until > NOW() THEN locked_until END AS locked_until
                     FROM users ORDER BY id"""
            )
            if not rows:
                print("[unlock] no accounts — the setup wizard at /setup is open")
                return 0
            print(f"{'id':>4}  {'username':<24} {'role':<8} {'fails':>5}  state")
            for row in rows:
                if row["locked"]:
                    state = "LOCKED (manual)"
                elif row["locked_until"]:
                    state = f"locked until {row['locked_until']:%Y-%m-%d %H:%M %Z}"
                else:
                    state = "ok"
                print(f"{row['id']:>4}  {row['username']:<24} {row['role']:<8} "
                      f"{row['failed_attempts']:>5}  {state}")
            return 0

        if args.user:
            result = await con.execute(
                """UPDATE users SET failed_attempts = 0, locked_until = NULL, locked = FALSE
                    WHERE username = $1""",
                args.user,
            )
            if result.endswith(" 0"):
                print(f"[unlock] no such account: {args.user}", file=sys.stderr)
                return 1
            print(f"[unlock] '{args.user}' unlocked")
            return 0

        result = await con.execute(
            """UPDATE users SET failed_attempts = 0, locked_until = NULL, locked = FALSE
                WHERE locked OR locked_until IS NOT NULL OR failed_attempts > 0"""
        )
        print(f"[unlock] {result.rsplit(' ', 1)[-1]} account(s) unlocked")
        return 0
    finally:
        await con.close()


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))
